src/jinja2/filters.py

```diff
@@ -218,6 +218,34 @@ def do_lower(s: str) -> str:
     return soft_str(s).lower()
 
 
+def do_pairs(value: t.Union[t.Mapping[K, V], Undefined]) -> t.Iterable[t.Tuple[K, V]]:
+    """Returns an iterator over the pairs of a mapping in the order that
+    the mapping's iterator returns.  In practical terms ``x|pairs`` is
+    the same as ``x.items()`` with two differences.  The first is that
+    if ``x`` is undefined an empty iterator is returned, secondly the
+    filter is a more portable version to alternative Jinja implementations
+    which might not have a ``.items()`` method on maps.
+
+    Example usage:
+
+    .. sourcecode:: html+jinja
+
+        <dl>
+        {% for key, value in my_dict|pairs %}
+            <dt>{{ key }}
+            <dd>{{ value }}
+        {% endfor %}
+        </dl>
+
+    .. versionadded:: 3.1
+    """
+    if isinstance(value, Undefined):
+        return
+    if not isinstance(value, abc.Mapping):
+        raise TypeError("Can only get pairs from mappings")
```

Comment:
It's weird to name this filter "pairs" if it only works with a mapping and not also with an iterable of pairs. Why not also accept iterables of pairs, or otherwise name this filter something else?

Reply:
@jab what would the behavior be for an iterable that already has pairs. return it unchanged? That seems to be a pretty pointless behavior.

Alternatively it could be named `|items` or `|entries` but I'm not sure if those are better names.

Reply:
If it's not a Mapping, falling back to yielding pairs would make the filter idempotent, which can be a nice property (in case .items() has already been called on the object), and maybe aid in composing with other filters?

Otherwise naming this "items" or "entries" sounds good to me.

Reply:
This requires the iterator to be active every single iteration and check if the items yielded are in fact pairs. And then if something goes wrong it just errors out in a new way. I just encourages an idempotent which I don't think is a good idea anyways.